### PR TITLE
create assessor contract

### PR DESCRIPTION
**Pull request by ilinzweilin**
- created assessor contract for asset valuation of tranches - important to determine current token value of each tranche. Tried to stick to the generic approach. The functions should work with more than 2 tranches.
- introduced a version of trancheManager (duplicate to Charly's pr). Didn't base on Charly's pr as it was still WIP. We will need to do some merge magic. Tried to use same interface and structure, but also added some additional functions.
- picked very descriptive names and left some comments for better understanding of the formulas used. Happy to change if not required.

refactoring:
- renamed some contract files to use camelCase like in the contract names
- removed desk mock and replaced by trancheManager

### CODE REVIEW

File: src/core/tranche/assessor.sol [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+         // losses: tranche debt can not be covered anymore.
+         // Tranche assets contain everything left in the loan pool + reserves of tranches with higher risk minus the outstanding debt of tranches with lower risk 
+         int assetValue = int(poolValue + equityReserve + trancheReserve - seniorDebt);
+         return (assetValue > 0) ? uint(assetValue) : 0;
+    }
+
+    // totalEquityReserve = sum of the reserves of all tranches with higher risk than the tranche with the index provided
+    function calcEquityReserve(uint startIndex) internal returns (uint) {
+        uint totalEquityReserve = 0;
+        for (uint i = startIndex; i < trancheManager.trancheCount(); i++) {
```
Comment (charleenfei) [inline src/core/tranche/assessor.sol]: i think this should be `trancheCount + 1` (index, not length)

File: src/core/tranche/assessor.sol [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+        trancheManager = TrancheManagerLike(trancheManager_);
+    }
+
+    function getAssetValueFor(address operator_) public returns (uint) {
+         OperatorLike operator = OperatorLike(operator_);
+         uint trancheDebt = operator.debt();
+         uint trancheReserve = operator.balance();
+         int trancheIndex = trancheManager.indexOf(operator_);
+
+         require(trancheIndex >= 0);
```
Comment (charleenfei) [inline src/core/tranche/assessor.sol]: maybe this `require` statement could come before the calls for the operator?

File: src/core/tranche/slicer.sol [hunk 3]
```diff
@@ -35,15 +35,22 @@ contract Slicer is DSNote {
...
-    constructor(address trancheManager_, address reserve_) public {
+    constructor(address assessor_, address reserve_) public {
         reserve = ReserveLike(reserve_);
-        trancheManager = TrancheManagerLike(trancheManager_);
+        assessor = AssessorLike(assessor_);
         wards[msg.sender] = 1;
     }
 
+    function file(bytes32 what, address address_) public note auth {
+        if (what == "operator") {
```
Comment (charleenfei) [inline src/core/tranche/slicer.sol]: i think contract dependencies should use the `depend` method, file is just for state variables

File: src/core/tranche/assessor.sol (hunk 1 above)
Comment (ilinzweilin) [reply src/core/tranche/assessor.sol]: trancheCount is the length. The last element in the array has the position -> length-1

File: src/core/tranche/assessor.sol (hunk 2 above)
Comment (ilinzweilin) [reply src/core/tranche/assessor.sol]: The call returns the trancheIndex

File: src/core/tranche/slicer.sol (hunk 3 above)
Comment (ilinzweilin) [reply src/core/tranche/slicer.sol]: this is a state variable, passed to the assessor as the trancheId

File: src/core/tranche/assessor.sol [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+
+    // Tranche.assets (Junior) = (Pool.value + Tranche.reserve - Senior.debt) > 0 && (Pool.value - Tranche.reserve - Senior.debt) || 0
+    function calcJuniorAssetValue(uint poolValue, uint trancheReserve, uint seniorDebt) internal returns (uint) {
+        int assetValue = int(poolValue + trancheReserve - seniorDebt);
+        return (assetValue > 0) ? uint(assetValue) : 0;
+    }
+
+    // Tranche.assets (Senior) = (Tranche.debt < (Pool.value + Junior.reserve)) && (Senior.debt + Tranche.reserve) || (Pool.value + Junior.reserve + Tranche.reserve)
+    function calcSeniorAssetValue(uint poolValue, uint trancheReserve, uint trancheDebt, uint juniorReserve) internal returns (uint) {
+        return ((poolValue + juniorReserve) >= trancheDebt) ? (trancheDebt + trancheReserve) : (poolValue + juniorReserve + trancheReserve);
```
Comment (manuelpolzhofer) [inline src/core/tranche/assessor.sol]: we calculate `poolValue+juniorReserve` twice. We could save the result
```javascript
uint currentMaxReturn = poolValue + juniorReserve;
return ((currentMaxReturn) >= trancheDebt) ? (trancheDebt + trancheReserve) : (currentMaxReturn + trancheReserve);
```

File: src/core/tranche/assessor.sol [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+    }
+
+    // Tranche.assets (Junior) = (Pool.value + Tranche.reserve - Senior.debt) > 0 && (Pool.value - Tranche.reserve - Senior.debt) || 0
+    function calcJuniorAssetValue(uint poolValue, uint trancheReserve, uint seniorDebt) internal returns (uint) {
+        int assetValue = int(poolValue + trancheReserve - seniorDebt);
+        return (assetValue > 0) ? uint(assetValue) : 0;
+    }
+
+    // Tranche.assets (Senior) = (Tranche.debt < (Pool.value + Junior.reserve)) && (Senior.debt + Tranche.reserve) || (Pool.value + Junior.reserve + Tranche.reserve)
+    function calcSeniorAssetValue(uint poolValue, uint trancheReserve, uint trancheDebt, uint juniorReserve) internal returns (uint) {
```
Comment (manuelpolzhofer) [inline src/core/tranche/assessor.sol]: maybe to be more specific
```suggestion
    function calcSeniorAssetValue(uint poolValue, uint seniorReserve, uint trancheDebt, uint juniorReserve) internal returns (uint) {
```

File: src/core/tranche/assessor.sol [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+            uint seniorDebt = calcSeniorDebt(); 
+            return calcJuniorAssetValue(poolValue, trancheReserve, seniorDebt);
+         }
+
+         uint juniorReserve = calcJuniorReserve();
+         return calcSeniorAssetValue(poolValue, trancheReserve, trancheDebt, juniorReserve);   
+    }
+
+    // Tranche.assets (Junior) = (Pool.value + Tranche.reserve - Senior.debt) > 0 && (Pool.value - Tranche.reserve - Senior.debt) || 0
+    function calcJuniorAssetValue(uint poolValue, uint trancheReserve, uint seniorDebt) internal returns (uint) {
```
Comment (manuelpolzhofer) [inline src/core/tranche/assessor.sol]:
```suggestion
    function calcJuniorAssetValue(uint poolValue, uint juniorReserve, uint seniorDebt) internal returns (uint) {
```

File: src/core/tranche/assessor.sol [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+    constructor(address trancheManager_) public {
+        trancheManager = TrancheManagerLike(trancheManager_);
+    }
+
+    function getAssetValueFor(address operator_) public returns (uint) {
+         OperatorLike operator = OperatorLike(operator_);
+         uint trancheDebt = operator.debt();
+         uint trancheReserve = operator.balance();
+         uint poolValue = trancheManager.poolValue();
+         if (trancheManager.isJunior(operator_)) {
```
Comment (manuelpolzhofer) [inline src/core/tranche/assessor.sol]: I think maybe the trancheManager shouldn't know that we are comparing operator addresses here.

why not just:
```javascript
if(operator_ == trancheManager.juniorTranche())
```

File: src/core/tranche/assessor.sol [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
+         return calcSeniorAssetValue(poolValue, trancheReserve, trancheDebt, juniorReserve);   
+    }
+
+    // Tranche.assets (Junior) = (Pool.value + Tranche.reserve - Senior.debt) > 0 && (Pool.value - Tranche.reserve - Senior.debt) || 0
+    function calcJuniorAssetValue(uint poolValue, uint trancheReserve, uint seniorDebt) internal returns (uint) {
+        int assetValue = int(poolValue + trancheReserve - seniorDebt);
+        return (assetValue > 0) ? uint(assetValue) : 0;
+    }
+
+    // Tranche.assets (Senior) = (Tranche.debt < (Pool.value + Junior.reserve)) && (Senior.debt + Tranche.reserve) || (Pool.value + Junior.reserve + Tranche.reserve)
```
Comment (manuelpolzhofer) [inline src/core/tranche/assessor.sol]: the comment is still a boolean expression which only results in `true` or `false`

**Comment by ilinzweilin** [comment]
opened new pr: https://github.com/centrifuge/tinlake/pull/152

**Pull request closed without merging**